src/charm.py

```diff
@@ -18,13 +20,83 @@ class WaltzOperatorCharm(charm.CharmBase):
     will be used by FINOS Waltz.
     """
 
+    _state = framework.StoredState()
+
     def __init__(self, *args):
         super().__init__(*args)
 
+        self._init_state()
+
+        # PostgreSQL relation hooks:
+        self.db = pgsql.PostgreSQLClient(self, "db")
+        self.framework.observe(
+            self.db.on.database_relation_joined, self._on_database_relation_joined
+        )
+        self.framework.observe(
+            self.db.on.database_relation_broken, self._on_database_relation_broken
+        )
+        self.framework.observe(self.db.on.master_changed, self._on_master_changed)
+
         # General hooks:
         self.framework.observe(self.on.waltz_pebble_ready, self._on_waltz_pebble_ready)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
 
+    def _init_state(self):
+        """Initializes the stored state."""
+        self._state.set_default(db_host=None)
+        self._state.set_default(db_port=None)
+        self._state.set_default(db_name=None)
+        self._state.set_default(db_username=None)
+        self._state.set_default(db_password=None)
```

Comment:
I've removed the StoredState. For credentials, I think we can remove them once we have that "integrator" charm, in case anyone wants to use their own postgresql database for Waltz. I've sent a PR tracking the removal here: https://github.com/pedroleaoc/waltz-integration-juju/pull/4

Reply:
I was going to challenge the idea of having a Postgres integrator because it didn't feel very Juju. People would have to 1. relate the charms and 2. 'juju config integrator'. That's double the work I would expect to pass DB credentials to a charm.

However, given the Waltz stack will be mostly deployed as bundle, users will only run 'juju config integrator cred=...'. So I am fine with the idea of an integrator charm. +, we will have just one interface mean which is nice.

Reply:
Are we sure waltz will always be deployed as bundle ? If the bundle will always include PostgresQL then what about institutions that already have PostgreSQL installations where they are not yet ready to migrate to the PostgreSQL charm ?

Reply:
That's where the integrator would come in...